crates/wasmtime/src/module/serialization.rs

```diff
@@ -0,0 +1,715 @@
+            MyCow::Owned(val) => val.serialize(dst),
+        }
+    }
+}
+
+impl<'a, 'b, T: Deserialize<'a>> Deserialize<'a> for MyCow<'b, T> {
+    fn deserialize<D>(src: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'a>,
+    {
+        Ok(MyCow::Owned(T::deserialize(src)?))
+    }
+}
+
+impl From<settings::OptLevel> for OptLevel {
+    fn from(level: settings::OptLevel) -> Self {
+        match level {
+            settings::OptLevel::Speed => OptLevel::Speed,
+            settings::OptLevel::SpeedAndSize => OptLevel::SpeedAndSize,
+            settings::OptLevel::None => OptLevel::None,
+        }
+    }
+}
+
+/// A small helper struct which defines modules are serialized.
+#[derive(Serialize, Deserialize)]
+struct SerializedModuleData<'a> {
+    /// All compiled artifacts needed by this module, where the last entry in
+    /// this list is the artifacts for the module itself.
+    artifacts: Vec<MyCow<'a, CompilationArtifacts>>,
+    /// Closed-over module values that are also needed for this module.
+    modules: Vec<SerializedModuleData<'a>>,
+    /// The index into the list of type tables that are used for this module's
+    /// type tables.
+    type_tables: usize,
+}
+
+impl<'a> SerializedModuleData<'a> {
+    pub fn new(module: &'a Module) -> (Self, Vec<MyCow<'a, TypeTables>>) {
+        let mut pushed = HashMap::new();
+        let mut tables = Vec::new();
+        return (module_data(module, &mut pushed, &mut tables), tables);
+
+        fn module_data<'a>(
+            module: &'a Module,
+            type_tables_pushed: &mut HashMap<usize, usize>,
+            type_tables: &mut Vec<MyCow<'a, TypeTables>>,
+        ) -> SerializedModuleData<'a> {
+            // Deduplicate `Arc<TypeTables>` using our two parameters to ensure we
+            // serialize type tables as little as possible.
+            let ptr = Arc::as_ptr(module.types());
+            let type_tables_idx = *type_tables_pushed.entry(ptr as usize).or_insert_with(|| {
+                type_tables.push(MyCow::Borrowed(module.types()));
+                type_tables.len() - 1
+            });
+            SerializedModuleData {
+                artifacts: module
+                    .inner
+                    .artifact_upvars
+                    .iter()
+                    .map(|i| MyCow::Borrowed(i.compilation_artifacts()))
+                    .chain(Some(MyCow::Borrowed(
+                        module.compiled_module().compilation_artifacts(),
+                    )))
+                    .collect(),
+                modules: module
+                    .inner
+                    .module_upvars
+                    .iter()
+                    .map(|i| module_data(i, type_tables_pushed, type_tables))
+                    .collect(),
+                type_tables: type_tables_idx,
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SerializedModule<'a> {
+    version: String,
```

Comment:
FWIW for this to be a bit more robust I think we'd want to move this outside of the bincode-encoded payload. Bincode may change formats (perhaps?) or we could change the structure here which means that if you're mismatching versions you're unlikely to get the error about different versions, that's only if the binary formats happen to align well.

Reply:
That makes sense to me.  I'll move this out and validate it from there.